lib/Onyx.js

```diff
@@ -1571,7 +1593,99 @@ function update(data) {
         }
     });
 
-    return clearPromise.then(() => Promise.all(_.map(promises, (p) => p())));
+    return clearPromise.then(() =>
+        multiGet(_.keys(updateQueue)).then((pairs) => {
+            const existingSnapshot = _.reduce(
+                pairs,
+                (acc, [key, value]) => {
+                    acc[key] = value;
+                    return acc;
+                },
+                {},
+            );
+
+            const updatedSnapshot = _.reduce(
+                updateQueue,
+                (acc, operations, key) => {
+                    // If the first operation is a set, we want to use that as the existing value.
+                    // Only the first operation can be a `set`, since when enqueuing the operations, we clear the whole queue each time a `set` is enqueued.
+                    const existingValue = operations[0].onyxMethod === METHOD.SET ? operations[0].value : existingSnapshot[key];
+
+                    // Remove the first operation if it was a `set`, since we already used it to get the existing value.
+                    const mergeOperations = _.filter(operations, ({onyxMethod}) => onyxMethod === METHOD.MERGE);
+
+                    // If there are no merge operations, it means only a single `set` operation was enqueued.
+                    // In this case, we should use the existing value.
+                    // Otherwise, apply the merge operations to the existing value.
+                    acc[key] = _.isEmpty(mergeOperations)
+                        ? existingValue
+                        : applyMerge(
+                              existingValue,
+                              _.map(mergeOperations, ({value}) => value),
+                              true,
+                          );
+                    return acc;
+                },
+                {},
+            );
+
+            const pairsForStorage = prepareKeyValuePairsForStorage(updatedSnapshot);
+
+            console.log('Onyx: data', data);
+            console.log('Onyx: updateQueue', updateQueue);
+            console.log('Onyx: existingSnapshot', existingSnapshot);
+            console.log('Onyx: updatedSnapshot', updatedSnapshot);
+
+            const updatesToBroadcast = [];
+            onyxCollectionKeySet.forEach((collectionKey) => {
+                const updatedCollection = _.reduce(
+                    _.keys(updatedSnapshot),
+                    (acc, key) => {
+                        if (isKeyMatch(collectionKey, key)) {
+                            acc[key] = updatedSnapshot[key];
+                            delete updatedSnapshot[key];
```

Comment:
A comment explaining why we delete here would be handy. It seems it's to prevent duplicate notification of subscribers, because `keysChanged` probably triggers individual key subscribers

Reply:
Yes, exactly. So the main idea is that we collect all the keys that belong to collections (even if they were updated individually), and process them as one collection while processing the non-collection keys as individuals.
The `keysChanged` notifies the subscribers subscribed to the whole collection as well as those subscribed to individual keys. So we remove the individual collection keys to avoid double notifications.